lms/djangoapps/certificates/signals.py

```diff
@@ -86,19 +87,58 @@ def _listen_for_passing_grade(sender, user, course_id, **kwargs):  # pylint: dis
 
     # Only SELF_PACED_ONLY flag enabled
     if waffle.waffle().is_enabled(waffle.SELF_PACED_ONLY):
-        if not courses.get_course_by_id(course_key, depth=0).self_paced:
+        if not courses.get_course_by_id(course_id, depth=0).self_paced:
             return
 
     # Only INSTRUCTOR_PACED_ONLY flag enabled
-    elif waffle.waffle().is_enabled(waffle.INSTRUCTOR_PACED_ONLY):
-        if courses.get_course_by_id(course_key, depth=0).self_paced:
+    if waffle.waffle().is_enabled(waffle.INSTRUCTOR_PACED_ONLY):
+        if courses.get_course_by_id(course_id, depth=0).self_paced:
             return
-    if GeneratedCertificate.certificate_for_student(self.user, self.course_id) is None:
-        generate_certificate.apply_async(
-            student=user,
-            course_key=course_id,
-        )
+    if fire_ungenerated_certificate_task(
+        user=user,
+        course_id=course_id
+    ):
         log.info(u'Certificate generation task initiated for {user} : {course} via passing grade'.format(
             user=user.id,
             course=course_id
         ))
+
+
+@receiver(LEARNER_NOW_VERIFIED, dispatch_uid="learner_track_changed")
+def _listen_for_track_change(sender, user, **kwargs):  # pylint: disable=unused-argument
+    """
+    Catches a track change signal, determines user status,
+    calls fire_ungenerated_certificate_task for passing grades
+    """
+    if (
+        not waffle.waffle().is_enabled(waffle.SELF_PACED_ONLY) and
+        not waffle.waffle().is_enabled(waffle.INSTRUCTOR_PACED_ONLY)
+    ):
+        return
+    # This will read a user grade, and if passed, will fire
```

Comment:
Not to be too anti-comment, but I think most of the words in this comment are made pretty clear by the words used in the code.

Actually, this is a compliment: the code is clear enough that no comment is needed.